src/modelValidator.js

```diff
@@ -1,357 +1,469 @@
+     * @property {string} code - Unique code to relate back to the error
+     * @property {string} message - Human readable message describing the problem
+     * @property {number} offset - Begin offset of the error
+     * @property {number} endOffset - End offset of the error
+     * @property {('ERROR'|'WARNING')} severity - Severity of the error
+     * @property {string|undefined} source - Text that relates to the cause of the error
+     * @property {Array|undefined} path - Path to the error
+     */
+
+    getDiagnostics() {
+        const actNameValidationErrors = this._checkIdentifiers('acts', 'act', /^<<.+>>$/)
+        const factNameValidationErrors = this._checkIdentifiers('facts', 'fact', /^\[.+\]$/)
+        const dutyNameValidationErrors = this._checkIdentifiers('duties', 'duty', /^<.+>$/)
 
-    if (operandsNode) {
-      for (let subNode of operandsNode.children) {
-        errors = errors.concat(this._validateParsedExpressionNode(subNode))
-      }
+        const referenceErrors = this._checkReferences()
+
+        return actNameValidationErrors.concat(factNameValidationErrors, dutyNameValidationErrors, referenceErrors)
     }
 
-    const operandNode = jsonc.findNodeAtLocation(expression, ['operand'])
+    /***
+     * Does a overall look at duplicate identifiers,
+     * e.g. <<act>> is not only findable on 'ACTS' field but as well
+     * on 'FACTS' field.
+     *
+     * @returns {ValidationError[]} Validation errors
+     * @private
+     */
+    _findOverallDuplicateIdentifiers() {
+        const validationError = [];
+        this.valuesOfIdentifiers.filter(value =>
+            this.multiIdentifierPaths[value].length > 1)
+            .forEach(key => {
+                const errors = this.multiIdentifierPaths[key]
+                    .map(path => {
+                        const node = jsonc.findNodeAtLocation(this.tree, path);
+                        const beginPosition = node.offset;
+                        const endPosition = node.offset + node.length;
+                        return new ValidationMessage(
+                            'LR0004',
+                            'Overall duplicate identifier',
+                            [beginPosition, endPosition],
+                            'ERROR',
+                            key.toString(),
+                            path
+                        )
+                    });
+                validationError.push(errors);
+            });
+        return validationError.flatMap(value => value);
+    }
 
-    if (operandNode) {
-      errors = errors.concat(this._validateParsedExpressionNode(operandNode.value))
+    /***
+     * Looks for duplicate identifiers only on the given field, e.g. only 'ACTS' field.
+     *
+     * @param flintItems identifier field
+     * @param flintItem identifier
+     * @returns {ValidationError[]} Validation errors
+     * @private
+     */
+    _findDuplicateIdentifiers(flintItems, flintItem) {
```

Comment:
Any duplicate that is found by this method is also found by the overall method. I'd drop the distinction and remove this function. The overall version looks very clean, and does less computation, and functionally there's very little difference, I think.

Reply:
Thanks for the feedback.

For a better picture of preferred result I would like to address some context about the methods and please correct me if Im wrong.

:: Indeed the methods do similar things but slightly differ when it comes to finding duplicates on identifiers.

**1st method: _findDuplicateIdentifiers(flintItems, flintItem)**
Looks only for duplicates for the given identifier, e.g. Acts. This means it wont look for duplicates on **Facts** and **Duties** or in between.

E.g. When **Acts** is given as flintItems, see img:
![first](https://user-images.githubusercontent.com/17723286/73952857-d4aa1c80-48ff-11ea-9219-e61c0055d2c2.png)

**2st method: _findOverallDuplicateIdentifiers()**
Looks for overall duplicates e.g. Acts, Facts, Duties and also in between. This means when there occurs a value "**test**" which occurs only once on Facts and Duties level it will tag those as duplicate as well.

E.g. see img:
![second](https://user-images.githubusercontent.com/17723286/73953403-a37e1c00-4900-11ea-88ff-35dc2299df7a.png)

If the second context is the preferred result than indeed the 1st method is unnecessary.

Reply:
Thanks for the clarification, this matches what I thought. Behaviourally both are fine. The second covers the first, so that's why I'd drop the first. 

In practice, the identifiers need to follow a pattern: `<<do something>>` for acts, `[some truth]` for facts and `<need to do something>` for duties. Therefore, if there are overall duplicates, one of these patterns is also violated. This means that in practice, these overall duplicates are less likely to occur, but still could happen. 

If I look at the code, the overall function is more readable, and more efficient. Therefore my recommendation to keep that one is based on that, not so much on the difference in functionality between the two.